File: java/cormen/algo1/src/chapter1/QuickSort.java
package chapter1;

public class QuickSort {
	
	private static void swap(int[] a, int i, int j) {
		int temp = a[i];
		a[i] = a[j];
		a[j] = temp;
	}
	
	private static int partition(int[] a, int low, int high) {
		int i = low - 1;
		int pivot = a[high];
		for (int j = low; j <= high; j++) {
			if (a[j] < pivot) {
				i++;
				swap(a, i, j);
			}
		}
		swap(a, i + 1, high);
		return i + 1;
	}
	
	public static void quickSort(int[] a, int low, int high) {
		if (low < high) {
			int pos = partition(a, low, high);
			quickSort(a, low, pos - 1);
			quickSort(a, pos + 1, high);
		}
	}

	public static void main(String[] args) {
		int a[] = {6, 2, 8, 3, 5, 4, 1, 9, 7};
		quickSort(a, 0, a.length - 1);
		SimpleArrayUtils.printSimpleArray(a);
	}
}

File: cpp/problems/cormen/quicksort.cpp
#include <iostream>

void swap(int &a, int &b) {
    int c = a;
    a = b;
    b = c;
}

int partition(int a[100], int low, int high) {
    int pivot = a[high];
    int k = low - 1;
    for(int i = low; i <= high; i++) {
        if (a[i] < pivot) {
            k++;
            swap(a[k], a[i]);
        }
    }
    swap(a[k + 1], a[high]);
    return k + 1;
}

void quicksort(int a[100], int low, int high) {
    if (low < high) {
        int mid = partition(a, low, high);
        quicksort(a, low, mid - 1);
        quicksort(a, mid + 1, high);
    }
}

void print(int a[100], int n) {
    for (int i = 0; i < n; i++) std::cout << a[i] << " ";
    std::cout << '\n';
}

int main(int argc, char *argv[]) {
    int a[100] = {9, 5, 2, 8, 6, 1, 3, 7, 4};
    quicksort(a, 0, 8);
    print(a, 9);
    return 0;
}
